fn.go

```diff
@@ -20,34 +34,200 @@ type Function struct {
+
+	metaApiVersion = "meta.gotemplating.fn.crossplane.io/v1alpha1"
+)
+
 // RunFunction runs the Function.
 func (f *Function) RunFunction(_ context.Context, req *fnv1beta1.RunFunctionRequest) (*fnv1beta1.RunFunctionResponse, error) {
 	f.log.Info("Running Function", "tag", req.GetMeta().GetTag())
 
-	// This creates a new response to the supplied request. Note that Functions
-	// are run in a pipeline! Other Functions may have run before this one. If
-	// they did, response.To will copy their desired state from req to rsp. Be
-	// sure to pass through any desired state your Function is not concerned
-	// with unmodified.
 	rsp := response.To(req, response.DefaultTTL)
 
-	// Input is supplied by the author of a Composition when they choose to run
-	// your Function. Input is arbitrary, except that it must be a KRM-like
-	// object. Supporting input is also optional - if you don't need to you can
-	// delete this, and delete the input directory.
 	in := &v1beta1.Input{}
 	if err := request.GetInput(req, in); err != nil {
 		response.Fatal(rsp, errors.Wrapf(err, "cannot get Function input from %T", req))
 		return rsp, nil
 	}
 
-	// TODO: Add your Function logic here!
-	//
-	// Take a look at function-sdk-go for some utilities for working with req
-	// and rsp - https://pkg.go.dev/github.com/crossplane/function-sdk-go
-	//
-	// Also, be sure to look at the tips in README.md
-	response.Normalf(rsp, "I was run with input %q", in.Example)
+	tg, err := NewTemplateSourceGetter(in)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, fmt.Sprintf(errFmtInvalidFunction, errCannotGet)))
+		return rsp, nil
+	}
+
+	tmpl, err := GetNewTemplateWithFunctionMaps().Parse(tg.GetTemplates())
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, fmt.Sprintf(errFmtInvalidFunction, errCannotParse)))
+		return rsp, nil
+	}
+
+	reqMap, err := convertToMap(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot convert request to map"))
+		return rsp, nil
+	}
+
+	f.log.Debug("constructed request map", "request", reqMap)
+
+	buf := &bytes.Buffer{}
+
+	if err := tmpl.Execute(buf, reqMap); err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot execute template"))
+		return rsp, nil
+	}
+
+	f.log.Debug("rendered manifests", "manifests", buf.String())
+
+	// Parse the rendered manifests.
+	var objs []*unstructured.Unstructured
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewBufferString(buf.String()), 1024)
+	for {
+		u := &unstructured.Unstructured{}
+		if err := decoder.Decode(&u); err != nil {
+			if err == io.EOF {
+				break
+			}
+			response.Fatal(rsp, errors.Wrap(err, "cannot decode manifest"))
+			return rsp, nil
+		}
+		if u != nil {
+			objs = append(objs, u)
+		}
+	}
+
+	// Get the desired composite resource from the request.
+	desiredComposite, err := request.GetDesiredCompositeResource(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot get desired composite resource"))
+		return rsp, nil
+	}
+
```

Comment:
AFAICT you're not getting the _observed_ composite resource anywhere. The function should create desired composed resources using the observed composite resource as an input.

Reply:
The actual logic which gets observed composite resource and creates desired composed resources is actually in the template [files](https://github.com/ezgidemirel/function-go-templating/blob/aeae407394468ddd39e9357d8278fd49ff59b2de/examples/composition.yaml#L21). 

But I should have used the observed composite resource to check if we want to update the composite resource status. Updated that part with the latest commit.